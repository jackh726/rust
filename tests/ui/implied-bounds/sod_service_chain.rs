// check-pass
// Found in a crater run on #118553

pub trait Debug {}

pub trait Service {
    type Input;
    type Output;
    type Error;
}

pub struct ServiceChain<P, S> {
    prev: P,
    service: S,
}
impl<P: Service, S: Service<Input = P::Output>> Service for ServiceChain<P, S>
where
    P::Error: 'static,
    S::Error: 'static,
{
    type Input = P::Input;
    type Output = S::Output;
    type Error = ();
}

pub struct ServiceChainBuilder<P: Service, S: Service<Input = P::Output>> {
    chain: ServiceChain<P, S>,
}
impl<P: Service, S: Service<Input = P::Output>> ServiceChainBuilder<P, S> {
    pub fn next<NS: Service<Input = S::Output>>(
    //~^ WARN method is missing
    //~| WARN this was previously accepted
        self,
    ) -> ServiceChainBuilder<ServiceChain<P, S>, NS> {
        panic!();
    }
}

fn main() {}
