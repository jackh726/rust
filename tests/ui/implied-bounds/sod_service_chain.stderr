warning: method is missing necessary lifetime bounds
  --> $DIR/sod_service_chain.rs:30:5
   |
LL | /     pub fn next<NS: Service<Input = S::Output>>(
LL | |
LL | |
LL | |         self,
LL | |     ) -> ServiceChainBuilder<ServiceChain<P, S>, NS> {
   | |____________________________________________________^
   |
   = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
   = note: for more information, see issue #109628 <https://github.com/rust-lang/rust/issues/109628>
   = note: to get more detailed errors, use `#[deny(implied_bounds_from_trait_impl)]`
   = note: `#[warn(implied_bounds_from_trait_impl)]` on by default

warning: 1 warning emitted

